flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDumpSerialization.java

```diff
@@ -124,50 +124,86 @@ public MetricSerializationResult serialize(
 			Map<Counter, Tuple2<QueryScopeInfo, String>> counters,
 			Map<Gauge<?>, Tuple2<QueryScopeInfo, String>> gauges,
 			Map<Histogram, Tuple2<QueryScopeInfo, String>> histograms,
-			Map<Meter, Tuple2<QueryScopeInfo, String>> meters) {
+			Map<Meter, Tuple2<QueryScopeInfo, String>> meters,
+			long maximumFramesize,
+			MetricQueryService queryService) {
 
 			buffer.clear();
+			boolean unregisterRemainingMetrics = false;
 
 			int numCounters = 0;
 			for (Map.Entry<Counter, Tuple2<QueryScopeInfo, String>> entry : counters.entrySet()) {
+				if (unregisterRemainingMetrics) {
+					queryService.unregister(entry.getKey());
+					continue;
+				}
+
 				try {
 					serializeCounter(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
 					numCounters++;
+					if (buffer.length() > maximumFramesize) {
+						unregisterRemainingMetrics = true;
```

Comment:
In this case a warning should be logged to inform users that not all metrics can be reporter to the UI.